src/coreclr/vm/methodtable.inl

```diff
@@ -1335,25 +1335,17 @@ inline OBJECTHANDLE MethodTable::GetLoaderAllocatorObjectHandle()
 }
 
 //==========================================================================================
-FORCEINLINE OBJECTREF MethodTable::GetManagedClassObjectIfExists()
+FORCEINLINE OBJECTREF MethodTable::GetPinnedManagedClassObjectIfExists()
 {
     LIMITED_METHOD_CONTRACT;
+    return GetRuntimeTypeObjectFromHandleFast(GetWriteableData_NoLogging()->GetExposedClassObjectHandle());
+}
 
-    // Logging will be done by the slow path
-    LOADERHANDLE handle = GetWriteableData_NoLogging()->GetExposedClassObjectHandle();
-
-    OBJECTREF retVal;
-
-    // GET_LOADERHANDLE_VALUE_FAST macro is inlined here to let us give hint to the compiler
-    // when the return value is not null.
-    if (!LoaderAllocator::GetHandleValueFast(handle, &retVal) &&
-        !GetLoaderAllocator()->GetHandleValueFastPhase2(handle, &retVal))
-    {
-        return NULL;
-    }
-
-    COMPILER_ASSUME(retVal != NULL);
-    return retVal;
+//==========================================================================================
+FORCEINLINE OBJECTREF MethodTable::GetManagedClassObjectIfExists()
```

Comment:
This still needs some fine-tunning. 

Run this:
```
using System.Diagnostics;
using System.Runtime.CompilerServices;

Stopwatch sw = new();
for (;;)
{
   sw.Restart();
   for (int i = 0; i < 1_000_000_000; i++) MyG<string>.M();
   Console.WriteLine(sw.ElapsedMilliseconds);
}

class MyG<T>
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static Type M() => typeof(T);
}
```
It has more than 2x regression compared to main with this change.

If you do this right, it should be a bit faster compared to main due to one less indirection.

Reply:
@jkotas Thanks, will take a look but I was hoping on Native PGO to kick in and reshape it

Reply:
I not sure whether native PGO would be able to clean this up completely. In any case, it would be unreliable at best.

Reply:
Dissassembly for `JIT_GetRuntimeType`
### Main (compiled with `/p:NoPgoOptimize=true`):
![image](https://user-images.githubusercontent.com/523221/191016935-dacbed23-48f4-4aa0-bda9-5dd543868024.png)

### This PR (also with `/p:NoPgoOptimize=true`):
![image](https://user-images.githubusercontent.com/523221/191017034-84059cbc-5557-4626-8a22-e75a3daa92b0.png)

Reply:
Experimenting with it to produce better asm now

Reply:
Pushed [a small change](https://github.com/dotnet/runtime/pull/75573/commits/e359e66fe0713eac59a6fb469247f3b1a3f0939c) that drastically improve asm for `JIT_GetRuntimeType` - I checked few times. Now idea why MSVC emitted such a bad code for the previous version.

![image](https://user-images.githubusercontent.com/523221/191021602-62b76bf4-3ffb-47f9-8151-faaad3337508.png)

Reply:
Ok, did some massage in C++ to produce better code here is what I ended up with: 
![image](https://user-images.githubusercontent.com/523221/191035497-a58228cd-2e55-40e2-a04c-ccf8a497e0a3.png)

Let me benchmark it on your benchmark

Reply:
![image](https://user-images.githubusercontent.com/523221/191036183-0406e40a-e2be-492b-bd3c-db35ffb1200a.png)

runtime-main - this PR, runtime7 - Main (the same commit PR is based on)
both were compiled with Release + `/p:NoPgoOptimize=true`

Disabled ReadyToRun to turn off managed static PGO just in case. Main is on noise-level faster but presumably Native PGO will make PR better as was highlighted on the screen ^

Reply:
> we have to do "test rdx, rdx"

We can keep the polarity of the low-bit to avoid this conditional jump.

> redundant "test rdx,rdx"

It is not redundant on some paths given the way the code is written. This should be fixable by structuring the code differently, similar to how it was structured before.

Reply:
> It is not redundant on some

Right, that's what I meant by "Doesn't MSVC support jump-threading, a compiler optimization that is supposed to handle such cases"

> We can keep the polarity of the low-bit to avoid this conditional jump.

@jkotas do you mean to set the lowest bit for gc ref instead? - is it safe to modify a bit in a gc reference like that?

Reply:
oh I see what you mean but then I would need to clear it before returning to JIT so we'll spend cycles there instead while currently I don't even need to clear it

Reply:
It is always better to spend a few extra instructions during JITing than to spend a few extra instruction when executing the JITed method.

It should work nicely together if you take https://github.com/dotnet/runtime/pull/75573#discussion_r974841009.